tests/integration/store/test_store_data.py

```diff
@@ -0,0 +1,161 @@
+    ]
+    data_types = list(
+        itertools.chain(*itertools.combinations(types, 2))
+    ) + [  # No combinations as it add much test time.
+        ("list-dict-1", [1, 2, {"data": [55, 66, 77], "dummy": "dum"}]),
+        ("list-dict-2", [1, 2, {"data": [111, 99, 88]}]),
+        ("dict-3", {"a": 1, "c": 1}),
+        ("dict-2", {"a": 1, "b": None}),
+    ]
+
+    app.layout = html.Div(
+        [
+            html.Div(id="output"),
+            html.Button("click", id="click"),
+            dcc.Store(id="store"),
+        ]
+    )
+
+    @app.callback(
+        Output("output", "children"),
+        [Input("store", "modified_timestamp")],
+        [State("store", "data")],
+    )
+    def on_data(ts, data):
+        if ts is None:
+            raise PreventUpdate
+        return json.dumps(data)
+
+    @app.callback(Output("store", "data"), [Input("click", "n_clicks")])
+    def on_click(n_clicks):
+        if n_clicks is None:
+            raise PreventUpdate
+        return data_types[n_clicks - 1][1]
+
+    dash_duo.start_server(app)
+
+    button = dash_duo.wait_for_element("#click")
+    for data_type in data_types:
+        button.click()
+        dash_duo.wait_for_text_to_equal("#output", json.dumps(data_type[1]))
+
+
+def test_stda002_nested_data(dash_duo):
+    app = dash.Dash(__name__)
+
+    nested = {"nested": {"nest": "much"}}
+    nested_list = dict(my_list=[1, 2, 3])
+
+    app.layout = html.Div(
+        [
+            dcc.Store(id="store", storage_type="local"),
+            html.Button("set object as key", id="obj-btn"),
+            html.Button("set list as key", id="list-btn"),
+            html.Output(id="output"),
+        ]
+    )
+
+    @app.callback(
+        Output("store", "data"),
+        [
+            Input("obj-btn", "n_clicks_timestamp"),
+            Input("list-btn", "n_clicks_timestamp"),
+        ],
+    )
+    def on_obj_click(obj_ts, list_ts):
+        if obj_ts is None and list_ts is None:
+            raise PreventUpdate
+
+        # python 3 got the default props bug. plotly/dash#396
+        if (obj_ts and not list_ts) or obj_ts > list_ts:
+            return nested
+        else:
+            return nested_list
+
+    @app.callback(
+        Output("output", "children"),
+        [Input("store", "modified_timestamp")],
+        [State("store", "data")],
+    )
+    def on_ts(ts, data):
+        if ts is None:
+            raise PreventUpdate
+        return json.dumps(data)
+
+    dash_duo.start_server(app)
+
+    obj_btn = dash_duo.wait_for_element("#obj-btn")
+    list_btn = dash_duo.find_element("#list-btn")
+
+    obj_btn.click()
+    dash_duo.wait_for_text_to_equal("#output", json.dumps(nested))
+    # it would of crashed the app before adding the recursive check.
+
+    list_btn.click()
+    dash_duo.wait_for_text_to_equal("#output", json.dumps(nested_list))
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 6),
+    reason="tests requires depedency only available in 3.6+",
+)
```

Comment:
makes a lot of sense. Pretty soon it will be time to move Percy over to py3 I'd think, so we don't have to choose between a modern helper package and taking snapshots...